include/kitty/ashenhurst_decomposition.hpp

```diff
@@ -0,0 +1,201 @@
+  Given functions f(.), g(.), and h(.) and a partition
+  on arguments into z and y. This function determines whether 
+  f(x) is decomposable into g(z, h(y)) where x = union(z,y) and 
+  intersect(z, y) = null.
+  This function does not check for permutation of variables given by
+  zs_index and ys_index. The elements in these vectors are treated as ordered
+  values.
+  \param tt The function to the check the decomposition on (function f)
+  \param zs_index The ordered set of indices of vector x (input to f) that
+                  are the inputs to outer_func (g).
+  \param ys_index The ordered set of indices of vector x (input to f) that are
+                  input to the inner_func (h).
+  \param outer_func The outer decomposition function (function g).
+  \param inner_func The inner decomposition function (function h).
+  \return true if the given decomposition is a valid one, false otherwise.
+*/
+template<class TTf, class TTg, class TTh>
+bool ashenhurst_decomposable( const TTf& tt,
+                              const std::vector<uint32_t> &zs_index,
+                              const std::vector<uint32_t> &ys_index,
+                              const TTg& outer_func,
+                              const TTh& inner_func )
+{
+  std::vector<TTf> y_vars;
+  std::vector<TTf> z_vars;
+
+  for ( const auto idx : ys_index )
+  {
+    auto var = tt.construct();
+    create_nth_var( var, idx );
+    y_vars.push_back( var );
+  }
+  for ( const auto idx : zs_index )
+  {
+    auto var = tt.construct();
+    create_nth_var( var, idx );
+    z_vars.push_back( var );
+  }
+  auto h = detail::compose_truth_table( inner_func, y_vars );
+  z_vars.push_back( h );
+  auto f = detail::compose_truth_table( outer_func, z_vars );
+  return equal( f, tt );
+}
+
+/*! Finds all of the possible decompositions of a function given an input 
+  partitioning.
+
+  \param tt The function to find all of its decompositions 
+  \param ys_index Indices indicating the partitioning of inputs
+  \param decomposition A vector of decomposition pairs. This serves as a return
+                       return container.
+  \return Returns the number of possible decompositions.
+*/
+template<class TTf, class TTg, class TTh>
+int ashenhurst_decomposition( const TTf& tt, const std::vector<uint32_t> &ys_index, std::vector<std::pair<TTg, TTh>>& decomposition )
+{
+  std::vector<uint32_t> zs_index = detail::enumerate_zs_index( ys_index, tt.num_vars() - 1 );
+  decomposition.clear();
+  //assert( decomposition.empty() );
+  int num_vars_h = ys_index.size();
+  int num_vars_g = tt.num_vars() - num_vars_h + 1;
+
+  uint64_t num_bits_h = ( uint64_t( 1 ) << num_vars_h );
+  uint64_t num_bits_g = ( uint64_t( 1 ) << num_vars_g );
+
+
+  // Note that this implementation does not support decompositions in which
+  // either f or h have more than 6 variables.
+  assert( num_bits_g <= 64 );
+  assert( num_bits_h <= 64 );
+
+  uint64_t num_funcs_h = ( uint64_t( 1 ) << num_bits_h );
```

Comment:
This code can be made much simpler by using `kitty::next_inplace`. Basically you start with a constant-0 truth table, and in a `do-while` loop call `next_inplace` on the truth table until this truth table is constant-0 again. Then you do not need `detail::make_binary_string` and also not `create_from_binary_string`. Even the loops become simpler because you do not need to count the size. Here is an example how it is used: https://github.com/msoeken/kitty/blob/bee07c6f15d8932009854216c208dbbb32db30fa/examples/npn_enumeration.cpp#L44-L54

Reply:
Great, I was looking for something like this but could not find it. I changed the loop to:

```cpp
  TTg g, g_contradiction;
  do {
    TTh h, h_contradiction;
    do {
      if (ashenhurst_decomposable(tt, zs_index, ys_index, g, h)) 
        decomposition.push_back(std::make_pair(g, h));
      next_inplace(h);
    }while (!equal(h, h_contradiction));
    next_inplace(g);
  } while (!equal(g, g_contradiction));
```
This is way better since the limitation on the number of variables is lifted.